File: web/scripts/architecture/ownership.mjs
import path from "node:path";
import ts from "typescript";
import { checkedText } from "./source-inputs.mjs";

const baseline = "82834bade1648da3067ebda1b1fc89c18577fd6a";

function exactKeys(value, fields) {
  return value && !Array.isArray(value) && Object.keys(value).sort().join(",") === [...fields].sort().join(",");
}

function safePath(value) {
  return typeof value === "string" && value.length > 0 && !value.includes("\\") &&
    !path.posix.isAbsolute(value) && value.split("/").every((part) => part !== "." && part !== ".." && part !== "");
}

export function ownershipIndex(registry) {
  if (!exactKeys(registry, ["schemaVersion", "baseline", "packages"]) || registry.schemaVersion !== 1 ||
    registry.baseline !== baseline || !Array.isArray(registry.packages) || registry.packages.length === 0) {
    throw new Error("architecture: invalid ownership registry");
  }
  const owners = new Map();
  const packages = new Set();
  for (const entry of registry.packages) {
    validatePackage(entry, packages);
    packages.add(entry.path);
    for (const file of entry.files) {
      validateFile(file, entry, owners);
      owners.set(file.path, { ...file, layer: entry.layer, module: entry.module, owner: entry.owner });
    }
  }
  return owners;
}

function validatePackage(entry, packages) {
  if (!exactKeys(entry, ["path", "layer", "module", "owner", "files"]) ||
    (entry.path !== "." && !safePath(entry.path)) || packages.has(entry.path) ||
    typeof entry.layer !== "string" || typeof entry.module !== "string" || !/^RF(?:0[1-9]|1[0-9]|2[0-2])$/.test(entry.owner) ||
    !Array.isArray(entry.files) || entry.files.length === 0) {
    throw new Error("architecture: invalid or duplicated package ownership");
  }
}

function validateFile(file, entry, owners) {
  if (!exactKeys(file, ["path", "kind"]) || !safePath(file.path) || path.posix.dirname(file.path) !== entry.path ||
    owners.has(file.path) || !["production", "test", "tool", "generated"].includes(file.kind)) {
    throw new Error("architecture: invalid or duplicated file ownership");
  }
}

export function readOwnership(root) {
  const input = checkedText(root, "quality/architecture/package-ownership.json");
  const registry = JSON.parse(input);
  const syntax = ts.parseJsonText("ownership.json", input);
  const visit = (node) => {
    if (ts.isObjectLiteralExpression(node)) {
      const keys = node.properties.map((property) => property.name.text);
      if (new Set(keys).size !== keys.length) {
        throw new Error("architecture: duplicate JSON key in ownership registry");
      }
    }
    ts.forEachChild(node, visit);
  };
  visit(syntax);
  ownershipIndex(registry);
  return registry;
}

File: internal/testkit/architecture/ownership.go
package architecture

import (
	"path"
	"slices"
	"strconv"
	"strings"
)

// OwnershipRegistry is an explicit file list, never a prefix exemption.
type OwnershipRegistry struct {
	SchemaVersion int                `json:"schemaVersion"`
	Baseline      string             `json:"baseline"`
	Packages      []PackageOwnership `json:"packages"`
}

// PackageOwnership assigns exactly one layer, module and refactoring owner.
type PackageOwnership struct {
	Path   string      `json:"path"`
	Layer  string      `json:"layer"`
	Module string      `json:"module"`
	Owner  string      `json:"owner"`
	Files  []OwnedFile `json:"files"`
}

// OwnedFile distinguishes test and tool sources without exempting their imports.
type OwnedFile struct {
	Path string `json:"path"`
	Kind string `json:"kind"`
}

// Violation identifies a failed rule and the dependency path that caused it.
type Violation struct {
	Rule            string   `json:"rule"`
	File            string   `json:"file"`
	Line            int      `json:"line"`
	Symbol          string   `json:"symbol"`
	DependencyChain []string `json:"dependencyChain"`
	Message         string   `json:"message"`
}

// LoadOwnership rejects unknown fields and multiple JSON documents.
func LoadOwnership(root string) (OwnershipRegistry, error) {
	data, err := readInventoryFile(root, "quality/architecture/package-ownership.json")
	if err != nil {
		return OwnershipRegistry{}, err
	}
	var registry OwnershipRegistry
	if err := decodeStrictJSON(data, &registry); err != nil {
		return OwnershipRegistry{}, err
	}
	return registry, nil
}

// ValidateOwnership compares exact source sets in both directions.
func ValidateOwnership(sources []string, registry OwnershipRegistry) []Violation {
	violations := make([]Violation, 0)
	if registry.SchemaVersion != 1 || !validCommitSHA(registry.Baseline) || len(sources) == 0 {
		return append(violations, ownershipViolation("", "invalid schema, baseline or empty source set"))
	}
	registered := make(map[string]bool)
	packages := make(map[string]bool)
	for _, entry := range registry.Packages {
		violations = append(violations, validateOwnershipEntry(entry, packages, registered)...)
	}
	actual := make(map[string]bool, len(sources))
	for _, name := range sources {
		actual[name] = true
		if !registered[name] {
			violations = append(violations, ownershipViolation(name, "source has no explicit owner"))
		}
	}
	for name := range registered {
		if !actual[name] {
			violations = append(violations, ownershipViolation(name, "registered source is absent"))
		}
	}
	slices.SortFunc(violations, compareViolations)
	return violations
}

func validateOwnershipEntry(entry PackageOwnership, packages, files map[string]bool) []Violation {
	violations := validatePhysicalLayer(entry)
	if !validOwnershipEntry(entry) || packages[entry.Path] {
		violations = append(violations, ownershipViolation(entry.Path, "invalid or duplicate package owner"))
	}
	packages[entry.Path] = true
	for _, file := range entry.Files {
		if !safeRepositoryPath(file.Path) || path.Dir(file.Path) != entry.Path || files[file.Path] {
			violations = append(violations, ownershipViolation(file.Path, "unsafe, duplicate or misplaced source"))
		}
		if !slices.Contains([]string{"production", "test", "tool", "generated"}, file.Kind) {
			violations = append(violations, ownershipViolation(file.Path, "invalid source kind"))
		}
		files[file.Path] = true
	}
	return violations
}

func validOwnershipEntry(entry PackageOwnership) bool {
	layers := []string{
		"cmd", "bootstrap", "model", "capability", "foundation", "service", "repo", "adapter", "transport",
		"testkit", "app", "feature", "component", "lib", "tool", "fixture", "contract", "documentation",
	}
	owner, err := strconv.Atoi(strings.TrimPrefix(entry.Owner, "RF"))
	return (entry.Path == "." || safeRepositoryPath(entry.Path)) &&
		slices.Contains(layers, entry.Layer) && entry.Module != "" &&
		len(entry.Owner) == 4 && strings.HasPrefix(entry.Owner, "RF") &&
		err == nil && owner >= 1 && owner <= 22 && len(entry.Files) > 0
}

func ownershipViolation(file, message string) Violation {
	return Violation{
		Rule: "GOV01", File: file, Line: 1, Symbol: "", DependencyChain: []string{file}, Message: message,
	}
}

func compareViolations(left, right Violation) int {
	if comparison := strings.Compare(left.File, right.File); comparison != 0 {
		return comparison
	}
	if comparison := strings.Compare(left.Rule, right.Rule); comparison != 0 {
		return comparison
	}
	return strings.Compare(left.Message, right.Message)
}
